tracker/enqueue_test.go

```diff
@@ -380,3 +380,261 @@ func TestBadObjectReferences(t *testing.T) {
+		}
+	}
+
+	// Starts getting called again
+	{
+		if err := trk.TrackReference(ref, thing2); err != nil {
+			t.Errorf("Track() = %v", err)
+		}
+		// New registrations should result in an immediate callback.
+		if got, want := calls, 4; got != want {
+			t.Errorf("Track() = %v, wanted %v", got, want)
+		}
+
+		trk.OnChanged(thing1)
+		if got, want := calls, 5; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// OnChanged non-accessor
+	{
+		// Check that passing in a resource that doesn't implement
+		// accessor won't panic.
+		trk.OnChanged("not an accessor")
+
+		if got, want := calls, 5; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// OnChanged non-accessor in DeletedFinalStateUnknown
+	{
+		// Check that passing in a DeletedFinalStateUnknown instance
+		// with a resource that doesn't implement accessor won't get
+		// Tracked called, and won't panic.
+		trk.OnChanged(cache.DeletedFinalStateUnknown{
+			Key: "ns/foo",
+			Obj: "not an accessor",
+		})
+
+		if got, want := calls, 5; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// Tracked gets called by DeletedFinalStateUnknown
+	{
+		trk.OnChanged(cache.DeletedFinalStateUnknown{
+			Key: "ns/foo",
+			Obj: thing1,
+		})
+		if got, want := calls, 6; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// Not called when something about the reference matching changes.
+	{
+		if err := trk.TrackReference(ref, thing2); err != nil {
+			t.Errorf("Track() = %v", err)
+		}
+		// New registrations should result in an immediate callback.
+		if got, want := calls, 6; got != want {
+			t.Errorf("Track() = %v, wanted %v", got, want)
+		}
+
+		thing1unlabeled := thing1.DeepCopy()
+		thing1unlabeled.Labels = nil
+		trk.OnChanged(thing1unlabeled)
+		if got, want := calls, 6; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+
+		thing1othernamespace := thing1.DeepCopy()
+		thing1othernamespace.Namespace = "another"
+		trk.OnChanged(thing1othernamespace)
+		if got, want := calls, 6; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+
+		thing1othergroup := thing1.DeepCopy()
+		thing1othergroup.APIVersion = "apps/v1"
+		trk.OnChanged(thing1othergroup)
+		if got, want := calls, 6; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+
+		thing1otherkind := thing1.DeepCopy()
+		thing1otherkind.Kind = "deployment"
+		trk.OnChanged(thing1otherkind)
+		if got, want := calls, 6; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+
+		// But with labels is still called
+		trk.OnChanged(thing1)
+		if got, want := calls, 7; got != want {
+			t.Errorf("OnChanged() = %v, wanted %v", got, want)
+		}
+	}
+
+	// Track bad object
+	{
+		if err := trk.TrackReference(ref, struct{}{}); err == nil {
+			t.Error("Track() = nil, wanted error")
+		}
+	}
```

Comment:
Your code permits matching by both paths now, i.e. cb will be invoked twice. I think you need a test for that case as well.

Reply:
Done